model-optimizer/extensions/front/onnx/loop_ext.py

```diff
@@ -130,21 +204,27 @@ def extract(cls, loop_node):
         #   1 .. loop_carried_dependencies_count - loop carried dependencies
         #   loop_carried_dependencies_count + 1 .. - scan outputs
 
-        body_graph.stage = 'front'
         # some of the inputs/outputs may not be connected but the normalization transformation will take care of it
         # connection Loop body nodes with external input edges
         next_loop_input_port_idx = sorted(loop_node.in_edges().keys())[-1] + 1
-        for (src_node, src_port), body_node, tensor_name in external_edges:
-            main_graph.add_edge(src_node, loop_node.id, **{'out': src_port,
-                                                           'in': next_loop_input_port_idx,
-                                                           'name': src_node,
-                                                           'fw_tensor_debug_info': [(src_node, tensor_name)],
-                                                           'in_attrs': ['in', 'name'],
-                                                           'out_attrs': ['out', 'name'],
-                                                           'data_attrs': ['fw_tensor_debug_info']}
-                                )
-            Loop.connect_body_input(loop_node, next_loop_input_port_idx, body_node)
-            next_loop_input_port_idx += 1
+        cur_graph = body_graph
+        for external_edges_subg in external_edges:
+            if 'parent_node' not in cur_graph.graph:
+                continue
+            cur_loop_node = cur_graph.graph['parent_node']
+            parent_graph = cur_loop_node.graph
+            for (src_node, src_port), body_node, tensor_name in external_edges_subg:
+                parent_graph.add_edge(src_node, cur_loop_node.id, **{'out': src_port,
+                                                               'in': next_loop_input_port_idx,
+                                                               'name': src_node,
+                                                               'fw_tensor_debug_info': [(src_node, tensor_name)],
+                                                               'in_attrs': ['in', 'name'],
+                                                               'out_attrs': ['out', 'name'],
+                                                               'data_attrs': ['fw_tensor_debug_info']}
+                                    )
+                Loop.connect_body_input(cur_loop_node, next_loop_input_port_idx, body_node)
+                next_loop_input_port_idx += 1
+            cur_graph = parent_graph
 
         # mark current iteration input Parameter node
```

Comment:
can it be a case when some inner Loop generates output that is a producer for some node from higher level outer graph (for example, grandpa graph)?

Reply:
Looks like no. We can't use variables from inner scope in outer without adding it to result